File: archive/leetcode-go/internal/code/002.go
package code

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var r *ListNode = nil
	var loop *ListNode = nil
	carry := 0
	remain := 0
	for l1 != nil || l2 != nil || carry > 0 {
		if l1 != nil {
			remain += l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			remain += l2.Val
			l2 = l2.Next
		}

		if remain > 9 {
			carry = remain / 10
			remain %= 10
		}
		cn := &ListNode{
			Next: nil,
			Val:  remain,
		}
		if r == nil {
			r = cn
		} else {
			loop.Next = cn
		}
		loop = cn
	}
	return r
}

File: leetcode/002.cpp
/**
 * @file   002.cpp
 * @brief  
 * @author zhenkai.sun
 * @email  zhenkai.sun@qq.com
 * @date   Wed Oct  1 05:43:26 AM CST 2025
 */

#include <algorithm>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
      ListNode head;
      ListNode *root = &head;
      int sm{0};
      while (l1 != nullptr || l2 != nullptr || sm > 0) {
        ListNode* nn = new ListNode(0, nullptr);
        if (l1) {
          sm += l1->val;
          l1 = l1->next;
        }
        if (l2) {
          sm += l2->val;
          l2 = l2->next;
        }

        nn->val = sm % 10;
        sm = sm / 10;

        root->next = nn;
        root = nn;
      }
      return head.next;
    }
};

int main() { return 0; }
